x/eth-bifrost/IMPLEMENTATION_SPEC.md

```diff
@@ -0,0 +1,133 @@
+
+## Signer component
+
+The signing apps sign transactions using secp256k1 such that the Ethereum smart contracts can verify them. The signing apps also have an ethereum address, because they have an identity in the Ethereum contract. They watch for new Ethereum-bound transactions using the ABCI app's query functionality, and submit their signatures in a transaction back to that ABCI.
+
+## Relayer component
+
+The relayer process is responsible for communication of state changes between Tendermint and Ethereum. It is stateless, and has at-least-once delivery semantics from one chain to another. If multiple relayers submit the same transaction to Ethereum and collide, they should get a nice error that informs them that the job is already done. If one relayer submits the same transaction to THORChain more than once, it should get a nice error informing it that the transaction has already been received. Then, every update delivered to either chain is idempotent.
+
+Generally anyone that wants their Bifrost to be successful has an incentive to run the relayer process.
+
+# Implementation steps
+ - Quick audit of solidity codebase for Peggy to determine what can be used and what needs to be fixed
+ - Fork of Peggy solidity codebase to be adapted into Bifrost Ethereum contract with additional needed changes
+ - Testnet (likely Ropsten) setup and deployment of Bifrost Ethereum contacts, and of some test ERC20 contracts, test ERC20 RUNE contract to be used for testing
+ - Quick audit of cosmos Oracle module to determine what can be used and what needs to be fixed
+ - Completion of Oracle module functionality
+ - Development of Bifrost module that accepts actions from the oracle module and mints tEth/tERC20 on THORChain and controls whitelisting
+ - Development of relayer service that watches Ethereum via Infura and submits transactions to the Oracle module
+ - Tests and QA of transfers from Ether/ERC20 into THORChain
+ - Enhance Bifrost module to be able to accept and burn tokens from users for sending back over the Bifrost to Ethereum
+ - Development of signer service that watches Bifrost module for these transactions, generates Ethereum signatures and submits them back to the Bifrost module for future relaying (or, to the reverse-oracle module if that path is explored)
+ - Enhance relayer to detect `m of n` signatures (or, to detect a `m of n` reverse-oracle state change if that path is explored) and relay them to the Ethereum smart contract
+ - Tests and QA of transfers of tEth/tERC20 on THORChain back to Ethereum
+ - Special rules for the official RUNE ERC20 contract to work differently to generic ERC20 contracts implemented
```

Comment:
Which rules do we need here?

Reply:
Good point, all tokens created on THORChain should probably locked rather than burned, right?